pkg/icingadb/ha.go

```diff
@@ -259,25 +275,41 @@ func (h *HA) realize(ctx context.Context, s *icingaredisv1.IcingaStatus, t *type
 			}
 
 			query := h.db.Rebind("SELECT id, heartbeat FROM icingadb_instance "+
-				"WHERE environment_id = ? AND responsible = ? AND id <> ? AND heartbeat > ?") + selectLock
```

Comment:
I haven't checked this thoroughly, but my first impression is that this can lead to unexpected rows when Icinga DB instances crash (repeatedly) in the `peerTimeout` timeframe. I think adding `ORDER BY heartbeat DESC` makes sense.

Reply:
Based on the takeover logic in Icinga DB, at most one node should be marked as `responsible` in the database. This selecting query, the upserting query setting this node's `responsible` field and the update query to mark everyone else as not `responsible` in case of a takeover is part of the same database transaction.

- https://github.com/Icinga/icingadb/blob/779afd1da35782cb164b3a296c201535bd4e170c/pkg/icingadb/ha.go#L281
- https://github.com/Icinga/icingadb/blob/779afd1da35782cb164b3a296c201535bd4e170c/pkg/icingadb/ha.go#L340
- https://github.com/Icinga/icingadb/blob/779afd1da35782cb164b3a296c201535bd4e170c/pkg/icingadb/ha.go#L346
- https://github.com/Icinga/icingadb/blob/779afd1da35782cb164b3a296c201535bd4e170c/pkg/icingadb/ha.go#L353

Thus, at most one row should be returned. Furthermore, for housekeeping there is the `removeOldInstances` method removing outdated entries.

While working on this PR, I have thought about opportunities with more then one `responsibility` Icinga DB instance and even temporary altered the code to allow working on multiple rows, but I was unable to enforce this behavior (unless I manually altered the database).

Reply:
Thanks for the extra work. You're right, it does look pretty solid.